Fix SPARK-698. From ExecutorRunner, launch java directly vs via the run scripts

cgrothaus opened the pull request:
Fixes https://spark-project.atlassian.net/browse/SPARK-698

Currently, doesn't handle "%FWDIR%conf\java-opts.cmd", as I found no docs about that.

Commit:
Fix SPARK-698. From ExecutorRunner, launch java directly instead via the run scripts.

mateiz commented on core/src/main/scala/spark/deploy/worker/ExecutorRunner.scala:
```diff
@@ -75,9 +75,45 @@ private[spark] class ExecutorRunner(
+        _javaLibPath,
+        "-Xms" + memory.toString + "M",
+        "-Xmx" + memory.toString + "M")
+    .filter(_ != "")
```
Why not just use `System.getenv` instead of `getEnvOrEmpty`, and then `filter(_ != null)`?

mateiz commented on core/src/main/scala/spark/deploy/worker/ExecutorRunner.scala:
```diff
@@ -75,9 +75,45 @@ private[spark] class ExecutorRunner(
+      "java"
+    } else {
+      getEnvOrEmpty("JAVA_HOME") + "/bin/java"
+    }
```
Here also you could just use `System.getenv` instead of `getEnvOrEmpty`

mateiz commented on run:
```diff
@@ -22,6 +22,8 @@ fi
   SPARK_MEM=${SPARK_DAEMON_MEMORY:-512m}
+  # Backup current SPARK_JAVA_OPTS for use in ExecutorRunner.scala
+  SPARK_NONDAEMON_JAVA_OPTS=$SPARK_JAVA_OPTS
   SPARK_JAVA_OPTS=$SPARK_DAEMON_JAVA_OPTS   # Empty by default
```
Instead of saving SPARK_NONDAEMON_JAVA_OPTS, I think it would be better to modify the scripts to use a different variable name for the JAVA_OPTS they do eventually use. For example:

```
OUR_JAVA_OPTS="$SPARK_JAVA_OPTS"
if [ "$1" = "spark.deploy.master.Master ]; then
    OUR_JAVA_OPTS="$SPARK_DAEMON_JAVA_OPTS"
fi
```

This will make the scripts clearer to understand.

mateiz commented:
Thanks a lot for the fix! I've made a few comments on the code but it looks good overall. Just update the pull request by pushing a new commit to your branch when you fix them.

Commit:
Incorporate feedback from mateiz:
- we do not need getEnvOrEmpty
- Instead of saving SPARK_NONDAEMON_JAVA_OPTS, it would be better to modify the scripts to use a different variable name for the JAVA_OPTS they do eventually use

cgrothaus commented:
Hello Matei, thanks for your feedback! I just updated the pull request.

mateiz commented:
Hey, sorry it's taken a while to merge this -- I wanted to test it to make sure the environment variables get passed the right way. I'm going to look at both it and #540 soon.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
I'm the Jenkins test bot for the UC, Berkeley AMPLab. I've noticed your pull request and will test it once an admin authorizes me to. Thanks for your submission!

AmplabJenkins commented:
I'm the Jenkins test bot for the UC, Berkeley AMPLab. I've noticed your pull request and will test it once an admin authorizes me to. Thanks for your submission!

mateiz commented:
Thanks for submitting this. I've now merged it into the master branch. I did one change, which is to read the classpath by calling an external script, because we can't be guaranteed that the ExecutorRunner thread will be launched by something that had the CLASSPATH environment variable set (specifically, in unit tests, when we use the "local-cluster" mode, that won't be true). I also made it split SPARK_JAVA_OPTS into separate arguments. See commits 15b00914c53f1f4f00a3313968f68a8f032e7cb7 and https://github.com/mesos/spark/commit/6c8d1b2ca618a1a17566ede46821c0807a1b11f5.
